File: P2/backend/src/main/java/com/jes/devlearn/domain/order/entity/Order.java
package com.jes.devlearn.domain.order.entity;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

@Entity
@Table(name = "orders")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EntityListeners(AuditingEntityListener.class)
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "order_no", unique = true, nullable = false, length = 32)
    private String orderNo;

    @Column(name = "user_id", nullable = false)
    private Long userId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private OrderStatus status = OrderStatus.PENDING;

    @Column(name = "total_amount", nullable = false)
    private Long totalAmount = 0L;

    @Column(name = "refunded_amount", nullable = false)
    private Long refundedAmount = 0L;

    @Column(name = "paid_at")
    private LocalDateTime paidAt;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    public Order(String orderNo, Long userId, Long totalAmount) {
        this.orderNo = orderNo;
        this.userId = userId;
        this.totalAmount = totalAmount;
        this.refundedAmount = 0L;
        this.status = OrderStatus.PENDING;
    }

    public boolean isOwnedBy(Long userId) {
        return this.userId.equals(userId);
    }

    public void markPaid() {
        this.status = OrderStatus.PAID;
        this.paidAt = LocalDateTime.now();
    }

    public void cancel() {
        this.status = OrderStatus.CANCELLED;
    }

    public void applyRefund(long amount, boolean fullyRefunded) {
        this.refundedAmount = (this.refundedAmount == null ? 0L : this.refundedAmount) + amount;
        this.status = fullyRefunded ? OrderStatus.REFUNDED : OrderStatus.PARTIAL_REFUNDED;
    }
}

File: P2/frontend/src/api/order.ts
import api from './index'

export type OrderStatus = 'PENDING' | 'PAID' | 'CANCELLED' | 'REFUNDED' | 'PARTIAL_REFUNDED'
export type OrderItemStatus = 'ACTIVE' | 'REFUNDED'

export interface OrderItem {
  id: number
  courseId: number
  courseTitle: string
  price: number
  status: OrderItemStatus
}

export interface Order {
  id: number
  orderNo: string
  status: OrderStatus
  totalAmount: number
  refundedAmount: number
  paidAt: string | null
  createdAt: string
  items: OrderItem[]
}

export const createOrder = async (): Promise<Order> => {
  const { data } = await api.post('/orders')
  return data.data
}

export const listMyOrders = async (status?: OrderStatus): Promise<Order[]> => {
  const { data } = await api.get('/orders', { params: status ? { status } : {} })
  return data.data
}

export const getOrder = async (id: number): Promise<Order> => {
  const { data } = await api.get(`/orders/${id}`)
  return data.data
}
